### ISSUE TITLE
Can't use typeof in declaration on imported type

### ISSUE THREAD

**Issue by AgentME** [facebook/flow#1691]
This seems like something that ought to work.

x.js

```js
/* @flow */

var X = {
  method() {return 123;}
};

export default X;
```

test.js

```js
/* @flow */

import type X from './x';

function f(x: typeof X) {
  return x.method();
}
```

```
$ flow
test.js:5
  5: function f(x: typeof X) {
                          ^ X. type referenced from value position
  3: import type X from './x';
                 ^ type X


Found 1 error
```

(Amusingly, if in test.js you forget to use `typeof`, then Flow's error message suggests that you might have forgot to use `typeof`, but when you do, you get this error.) 

Related: https://github.com/facebook/flow/issues/1189. Both situations are caused by imported types being disallowed inside non-Flow / "value position" syntax nested inside of a type declaration.

**Comment by samwgoldman** [comment]
Isn't this expected behavior? And isn't the error message pretty specific about why this is wrong? `typeof x` returns the type of `x` where `x` is some value in scope. `import type` introduces a type binding (not a value binding).

Maybe you meant to write `import X from './X'` instead, which introduces a value binding which you can then use with `typeof`? There's also an `import typeof` which is the same as `import X from ...` followed by `typeof X`.

**Comment by AgentME** [comment]
I want './x' not to be imported at runtime. We have a lot of circular type dependencies and some files containing types that have side effects when executed, and things don't always go well if we have all of these be regular imports that get executed immediately/ever.

I had been under the impression that `import type` works both as a way to import `export type` exports, and as a way to do regular imports in a way that only Flow sees for use in type declarations that doesn't cause the file to be imported at runtime. Reading the documentation now (I'd been using `import type` since before it was documented), I see there's a note about that this specific case doesn't work, but it feels pretty arbitrary compared to my previous understanding of `import type`. It seems a little strange that `import typeof` is called a shorthand when it does something that's otherwise impossible (doing it without an import seen at runtime).

`import typeof` does work for my case here, though I can't shake the impression that `import type` as I see it now works is arbitrarily limited, that this problem is a subset of https://github.com/facebook/flow/issues/1189, and `import typeof` is a special-case solution that misses the general problem.

**Comment by jeffmo** [comment]
`import type {foo} from "mymodule"` is sugar for `import {foo as foo_tmp} from "mymodule"; type foo = foo_tmp;` *except* that, because it's condensed into a single statement, we can forgo the runtime effects of the `import`.

So keeping this in mind, doing something like this:

```javascript
import {foo as foo_tmp} from "mymodule";
type foo = foo_tmp;
```

Only makes sense if `foo` is a type. If `foo` were a value (like a number/string/etc), this wouldn't make sense for the same reason that a type alias of a runtime value doesn't make sense:

```javascript
var a = 42;
type foo = a; // Error: Can't make a type-alias of a non-type.
```

There is exactly one kind of runtime value that is also a type: Classes:

```javascript
class Foo {}
type foo = Foo; // type of instances of `Foo`
```

`typeof` only works with variables that are runtime values:

```javascript
var a = 42;
var b: typeof a; // the "type of" a is a number

type foo = number;
var c: typeof foo; // Error: The "type of" a type doesn't make much sense
```

So with these baseline rules, it follows that `import type` can only work on exported types. `import` and `import typeof` can only work on exported runtime values.

**Comment by AgentME** [comment]
Thanks, this makes more sense now.

>Only makes sense if foo is a type. If foo were a value (like a number/string/etc), this wouldn't make sense for the same reason that a type alias of a runtime value doesn't make sense:

Huh, I didn't know this wasn't allowed. I've been using `import type` to import non-class variables a lot (usually just as a stand-in to show that it's meant to be the type of a class defined using an old pre-ES6 framework's class definer function in files we haven't fully flow-typed yet) and Flow never complained about that, so I got the wrong impression.

If this gives an error:
```js
/* @flow */
var a = 42;
type foo = a; // Error: Can't make a type-alias of a non-type.
```

Then I think it makes sense that this should too:
```js
/* @flow */
var a = 42;
export default a;
```

```js
/* @flow */
import type a from './a';
```

**Comment by jeffmo** [comment]
>  I've been using `import type` to import non-class variables a lot
> ...
> usually just as a stand-in to show that it's meant to be the type of a class defined using an old pre-ES6 framework's class definer function in files we haven't fully flow-typed yet

Ah, I bet it works because non-@flow files are typed as `any` (so Flow wouldn't know if the file actually has a class export or not -- hence the lack of an error).

I added errors for importing like this from @flow-files back in 0.22 though. For example, here's what happens with your example using Flow 0.23: https://gist.github.com/jeffmo/c7a37142aea486b69ed412098f89737d

**Comment by AgentME** [comment]
Oh, sorry! I was testing with Flow as I wrote that, but I failed to post the exact example I was using. Change the `42` to `{}` and you won't get an error when `import type` is used. I get the same error as you when I actually try with `42`.

**Comment by jeffmo** [comment]
Aha -- yes you've found an [easter egg](https://github.com/facebook/flow/blob/master/src/typing/flow_js.ml#L999-L1006) :)

tldr: It only works with object types because people like to do things like the following:

```javascript
import type Immutable from "immutable";
var a: Immutable.Map = ...
```
rather than
```javascript
import type {Map as ImmutableMap} from "immutable";
var a: ImmutableMap = ...
```

The latter would be more correct, but it's hard to argue with how convenient the former is -- so I didn't want to take it away until I have a safer alternative to offer. The first example also suffers from the same subtle value-into-type-space propagation issues that could exist before when `import type` would *not* error on non-type exports.

My safer future alternative is to introduce `import type * as Immutable from "immutable";` which will provide a special kind of object whose properties are only the type-compatible exports (types/classes/interfaces). Because this will be a special kind of object with only type properties, we don't have to worry about any subtle propagation bugs.

Anyway -- carry on taking advantage of this until `import type * as` comes along. When it does I'll be sure the change is documented clearly in both the Changelog and the new error message

**Comment by AgentME** [comment]
Thanks, that makes sense!
